core/gateway.go

```diff
@@ -239,13 +241,34 @@ func wrapSolveError(inputErr *error, gw bkgw.Client) {
 		if err := proc.Wait(); err != nil {
 			return
 		}
-		stdout := strings.TrimSpace(ctrOut.String())
-		stderr := strings.TrimSpace(ctrErr.String())
-		returnErr = fmt.Errorf("%w\nStdout:\n%s\nStderr:\n%s", returnErr, stdout, stderr)
+
+		returnErr = &ExecError{
+			original: returnErr,
+			Cmd:      execOp.Exec.Meta.Args,
+			// FIXME: extracting the exit code from the error message is a
+			// quick hack.
+			ExitCode: parseExitCode(returnErr.Error()),
```

Comment:
I think one other option would be something like this:
1. Instead of returning `0` in [the shim here](https://github.com/sipsma/dagger/blob/b3517009376241d5b78aa96e98231d8924317670/cmd/shim/main.go#L252-L252), return the value of the actual original exit code (as was written [here](https://github.com/sipsma/dagger/blob/b3517009376241d5b78aa96e98231d8924317670/cmd/shim/main.go#L305-L305) at the end of the failed exec).
2. I *think* that probably results in [an error being set here](https://github.com/sipsma/dagger/blob/b3517009376241d5b78aa96e98231d8924317670/core/gateway.go#L239-L239). And I also *think* it's probably of [this `ExitError` type](https://github.com/sipsma/buildkit/blob/dd53dca59c951b7b9a0f13815ca7477c37f4a6c2/frontend/gateway/pb/exit.go#L26-L26). If so, then you should be able to get the exit code direct from there

If that doesn't work out, there's other ways probably, but it's not a huge deal, can be fixed in the future.

Reply:
Yeah totally. Haven't even looked into it, I just wanted to get it out the door, so leaving for last 😅

Reply:
@sipsma, your tip on the [ExitError type](https://github.com/sipsma/buildkit/blob/dd53dca59c951b7b9a0f13815ca7477c37f4a6c2/frontend/gateway/pb/exit.go#L26-L26) was very useful, thanks!